hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMRatisServerImpl.java

```diff
@@ -227,12 +240,62 @@ public RaftPeerId getPeerId() {
               "localHost: {}, OZONE_SCM_NAMES: {}, selfPeerId: {}",
           localHost, conf.get(ScmConfigKeys.OZONE_SCM_NAMES), selfPeerId);
 
-      raftGroupId = RaftGroupId.valueOf(UUID.nameUUIDFromBytes(
-          SCM_SERVICE_ID.getBytes(StandardCharsets.UTF_8)));
+      String groupId = clusterId == null ? SCM_SERVICE_ID : clusterId;
+      raftGroupId = RaftGroupId.valueOf(
+          UUID.nameUUIDFromBytes(groupId.getBytes(StandardCharsets.UTF_8)));
 
       raftGroup = RaftGroup.valueOf(raftGroupId, raftPeers);
     }
 
+    SCMHAGroupBuilder(final SCMHANodeDetails details, final String clusterId) {
+
+      Preconditions.checkNotNull(clusterId);
+      // RaftGroupId is the clusterId
+      raftGroupId = RaftGroupId.valueOf(buildRaftGroupId(clusterId));
+      final String scmNodeId = details.getLocalNodeDetails().getNodeId();
+      Preconditions.checkNotNull(scmNodeId);
+      selfPeerId = RaftPeerId.getRaftPeerId(scmNodeId);
+      Preconditions.checkNotNull(scmNodeId);
+      SCMNodeDetails localNodeDetails = details.getLocalNodeDetails();
+
+      InetSocketAddress ratisAddr =
+          new InetSocketAddress(localNodeDetails.getInetAddress(),
+              localNodeDetails.getRatisPort());
+
+      RaftPeer localRaftPeer =
+          RaftPeer.newBuilder().setId(selfPeerId).setAddress(ratisAddr)
+              .build();
+
+      List<RaftPeer> raftPeers = new ArrayList<>();
+      // Add this Ratis server to the Ratis ring
+      raftPeers.add(localRaftPeer);
+
+      for (SCMNodeDetails peerInfo : details.getPeerNodeDetails()) {
+        String peerNodeId = peerInfo.getNodeId();
+        RaftPeerId raftPeerId = RaftPeerId.valueOf(peerNodeId);
+        RaftPeer raftPeer;
+        if (peerInfo.isHostUnresolved()) {
+          raftPeer = RaftPeer.newBuilder().setId(raftPeerId)
```

Comment:
Question: When the host is unresolved, we still construct raftPeer, do we fail here if it might be due to incorrect hostname. I see OM also followed similar approach, just trying to understand reason.

Reply:
Its following OM model.